Add BIP152 (Compact Blocks) structures

stevenroose opened the pull request:
Adds the basic structures for BIP152 and a method to create a compact block from a full block.

Depends on https://github.com/rust-bitcoin/bitcoin_hashes/pull/46.

tamasblummer commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,292 @@
+                .map(|tx| {
+                    ShortId::calculate_with_siphash_keys(
+                        &match use_wtxid {
+                            true => tx.bitcoin_hash(),
```
SendCmpct version and this should be better linked somehow.

stevenroose replied:
I think he is referring to the fact that the `use_wtxid` should be linked to the `version` field (1 -> false; 2 -> true) somehow. I'll change that to passing the version number and return an Error if it's not 1 or 2.

stevenroose commented:
So I understand we don't want a `siphasher` dependency, but rather a `bitcoin_hashes::siphash` impl?

tamasblummer commented:
yes, siphash is also used in BIP158. It would be better included into bitcoin_hashes

nothingmuch commented on src/util/bip152.rs:
```diff
@@ -79,6 +137,39 @@ pub struct HeaderAndShortIds {
+            header: block.header.clone(),
+            nonce: nonce,
+            // Provide coinbase prefilled.
+            prefilled_txs: vec![PrefilledTransaction(0, block.txdata[0].clone())],
```
This seems like hard coding a specific scenario for block transmission, but I'm not sure that's really a meaningful limitation in practice.

Perhaps a variant of `from_block` is warranted, with a parameter for specifying which transactions to prefill. I tried to do this based on a `Fn(usize, &Transaction) -> bool` argument, and also considered specifying which indexes to prefill, but I couldn't find a simple enough solution to make it seem justifiable.

I can think of two use cases where doing this could eliminate a block transactions request, both of which seem a bit contrived:

1. an edge node that relays txs and blocks to whitelisted, non-public nodes, and keeps track of which transactions each node knows about.
2. mining setting, where the miner knows some txs haven't been broadcast yet and wants to ensure their block can be validated without additional round trips

stevenroose replied:
Can you make a clear summary of what functionality you are missing?

I have to admit I built the interfacing functions based on some RPC functionality we use in Elements. (We added some stateless RPC calls to use the C++ CompactBlocks code, which we are calling from a Rust program, so I ported them to native Rust instead.)

stevenroose replied:
I'm changing `from_block` to a more general form taking a `&[bool]` to indicate which txs to prefill.

stevenroose commented on Cargo.toml:
```diff
@@ -49,3 +49,7 @@ version = "=0.3.2"
 features = [ "rand" ]
+
+[dependencies.siphasher]
+version = "0.3.0"
```
https://github.com/rust-bitcoin/bitcoin_hashes/pull/46

stevenroose commented:
> yes, siphash is also used in BIP158. It would be better included into bitcoin_hashes

@tamasblummer 
I added a siphash24 module in bitcoin_hashes: https://github.com/rust-bitcoin/bitcoin_hashes/pull/46

Could you check if that suits your need for BIP 158 usage? I'm actually very interested in BIP-158 implementation in rust-bitcoin. Are you planning to do them?

tamasblummer commented:
@stevenroose will review tomorrow. I already implemented and using BIP158 in murmel: https://github.com/rust-bitcoin/murmel/blob/master/src/blockfilter.rs

stevenroose commented:
> I already implemented and using BIP158 in murmel: https://github.com/rust-bitcoin/murmel/blob/master/src/blockfilter.rs

@tamasblummer would you mind upstreaming that to rust-bitcoin? Into `src/util/bip185.rs` f.e.?

stevenroose commented:
@tamasblummer I made 2 changes:
- `HeaderAndShortIds::from_block` now takes a version number instead of the `use_wtxid` bool. It returns an error if the version is not 1 or 2. This is probably better for backwards compatibility when we add support for other versions.
- `HeaderAndShortIds::from_block` now takes an additional `&[bool]` argument to indicate which transactions to prefill. It must be of length one less than the nb of txs in the block because the coinbase is always prefilled. This is quite rudimentary support for arbitrary fillings. I thought about a closure argument `Fn(&sha256d::Hash) -> bool`. One could pass a `HashSet::contains` or `HashMap::contains_key` into that argument, f.e., as well as a custom method. Would that be a better solution? Thought? @apoelstra @dongcarl

stevenroose commented:
This is rebased and ready to be reviewed. Uses the new bitcoin_hashes siphash impl.

TheBlueMatt commented on src/network/message_compact_blocks.rs:
```diff
@@ -0,0 +1,39 @@
+}
+impl_consensus_encoding!(SendCmpct, send_compact, version);
+
+/// cmpctblock message
```
Can you add add a warning to the effect of "note that the rules for validation before relay for compact blocks is different from headers and regular block messages. Thus, you cannot use compact blocks when relying on an upstream full node to have validated data being forwarded to you".

stevenroose replied:
Added.

stevenroose commented:
rebased

stevenroose commented:
fixed silly rebasing issue

stevenroose commented:
This PR is passing the tests (except for the one that [also fails on master](https://travis-ci.org/rust-bitcoin/rust-bitcoin/builds/593127075)).

apoelstra commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,335 @@
+/// A PrefilledTransaction structure is used in HeaderAndShortIDs to
+/// provide a list of a few transactions explicitly.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct PrefilledTransaction(pub u64, pub Transaction);
```
May be more efficient for users for this to hold a reference to the `Transaction`.

stevenroose replied:
This optimization is only possible in the case where you're encoding a compact block, but (1) those are usually transported out of the context so they can't have lifetimes of the original block and (2) by default only the coinbase tx is prefilled and in most cases the nb of prefilled txs will be minimal.

apoelstra commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,335 @@
+            block_hash: sha256d::Hash::consensus_decode(&mut d)?,
+            indexes: {
+                // Manually decode indexes because they are differentially encoded VarInts.
+                let differential: Vec<VarInt> = Decodable::consensus_decode(&mut d)?;
```
Not going to block this PR on it, but this could be done much more efficiently by avoiding the temporary `Vec`.

stevenroose replied:
Done in separate commit.

stevenroose commented:
K I rebased this and updated it to use the new hash newtypes for BlockHash. In one method it expects a "txid" that can be either `Txid` or `Wtxid`, so I parametrized that as `T: AsRef<[u8]>`. Is that to broad? I think it's fine.

TheBlueMatt commented:
Looks like the previous travis failure was a legitimate addition-overflow crash:

```
thread 'tests::duplicate_crash' panicked at 'attempt to add with overflow', /home/matt/Documents/Projects/Bitcoin/rust-bitcoin/src/util/bip152.rs:284:21
stack backtrace:
...
  14: core::panicking::panic
             at src/libcore/panicking.rs:52
  15: <bitcoin::util::bip152::BlockTransactionsRequest as bitcoin::consensus::encode::Decodable>::consensus_decode
             at /home/matt/Documents/Projects/Bitcoin/rust-bitcoin/src/util/bip152.rs:284
  16: <bitcoin::network::message_compact_blocks::GetBlockTxn as bitcoin::consensus::encode::Decodable>::consensus_decode
             at /home/matt/Documents/Projects/Bitcoin/rust-bitcoin/src/internal_macros.rs:39
  17: <bitcoin::network::message::RawNetworkMessage as bitcoin::consensus::encode::Decodable>::consensus_decode
             at /home/matt/Documents/Projects/Bitcoin/rust-bitcoin/src/network/message.rs:340
  18: bitcoin::consensus::encode::deserialize_partial
             at /home/matt/Documents/Projects/Bitcoin/rust-bitcoin/src/consensus/encode.rs:184
  19: bitcoin::consensus::encode::deserialize
             at /home/matt/Documents/Projects/Bitcoin/rust-bitcoin/src/consensus/encode.rs:168
  20: deser_net_msg::do_test
             at fuzz_targets/deser_net_msg.rs:4
```

stevenroose commented:
Yeah there was a potential overflow. Amazing!

Fixed that now, also verified that the specific fuzz input no longer causes a panic.

stevenroose commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,353 @@
+/// A PrefilledTransaction structure is used in HeaderAndShortIDs to
+/// provide a list of a few transactions explicitly.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct PrefilledTransaction(pub u64, pub Transaction);
```
ACK. Changed (pushing later).

stevenroose commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,353 @@
+    pub fn calculate_siphash_keys(header: &BlockHeader, nonce: u64) -> (u64, u64) {
+        // 1. single-SHA256 hashing the block header with the nonce appended (in little-endian)
+        let h = {
+            let mut b: Vec<u8> = vec![];
```
Whoops. Yeah this is an old MR. I recently changes all the `Hash::hash` usage into engine usage but not in this PR yet. ACK obviously.

stevenroose commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,353 @@
+
+        // 3. Dropping the 2 most significant bytes from the SipHash output to make it 6 bytes.
+        let mut id = ShortId([0; 6]);
+        id.0.copy_from_slice(&hash[0..6]);
```
Yeah I looked for an equivalent unit test in Core and couldn't find one. It doesn't make quite sense to add a unit test if the only way you can find test vectors is by using your own code.. I could try dig into Core's unit tests and add a test there and then reuse the values..

stevenroose replied:
Hmm, so Greg Sanders implemented compact blocks in Elements a few years ago where I got that test vector from. (Core didn't have support yet at the time.) The vector test breaks with `[2..]`, so I'm gonna have to get a vector from Core, hoping that's possible atm.

jrawsthorne replied:
That's weird. I've used this in https://github.com/jrawsthorne/rust-bitcoin-node and it works correctly with core

sgeisler replied:
As there's now a working test vector (I hope Core and Elements are actually that close in this regard) I guess we can assume this to be ok.

stevenroose commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,353 @@
+    ///
+    /// The [prefill] slice indicates which transactions should be prefilled in
+    /// the block. It should contain the indexes in the block of the txs to
+    /// prefill. It must be in ordered. 0 should not be included as the
```
Ah yeah that's a typo.

stevenroose commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,353 @@
+            };
+
+            if prefill_tx {
+                let diff_idx = idx - last_prefill;
+                last_prefill = idx;
```
Fixed.

stevenroose commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,353 @@
+        len += VarInt(self.indexes.len() as u64).consensus_encode(&mut s)?;
+        let mut last_idx = 0;
+        for idx in &self.indexes {
+            len += VarInt(*idx - last_idx).consensus_encode(&mut s)?;
```
ack

stevenroose commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,353 @@
+            indexes: {
+                // Manually decode indexes because they are differentially encoded VarInts.
+                let nb_indexes = VarInt::consensus_decode(&mut d)?.0 as usize;
+                let mut indexes = Vec::with_capacity(nb_indexes);
```
ack

sgeisler replied:
Did you add a safeguard somewhere?

stevenroose replied:
Whoops, did now!

stevenroose commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,353 @@
+                let mut last_index: u64 = 0;
+                for _ in 0..nb_indexes {
+                    let differential: VarInt = Decodable::consensus_decode(&mut d)?;
+                    last_index  = match last_index.checked_add(differential.0) {
```
ack

stevenroose commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,353 @@
+                ));
+            }
+        }
+
```
Good point. I added
```rust
    /// The prefill slice provided was invalid.
    InvalidPrefill,
```

stevenroose commented on src/network/message_compact_blocks.rs:
```diff
@@ -0,0 +1,44 @@
+    /// Request to be send compact blocks.
+    pub send_compact: bool,
+    /// Compact Blocks protocol version number.
+    pub version: u32,
```
Good find!

stevenroose commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,353 @@
+    }
+}
+
+/// A BlockTransactionsRequest structure is used to list transaction indexes
```
Hmm, I usually do square brackets because that (supposedly) allows rustdocs to make autolinks.

sgeisler replied:
Ah, I don't have a problem with that either, links are always better. I just saw a few occurrences where there were none. While this could be seen as  valid if you reference the data structure from the BIP instead of our types it seems like a useless distinction.

stevenroose replied:
Yeah my Rust coding style changed a bit over the last year :sweat_smile: I changed them all to square brackets in the entire module.

stevenroose commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,353 @@
+
+#[cfg(test)]
+mod test {
+    use super::*;
```
Yeah I'm thinking the same. That will be for tomorrow, though.

sgeisler replied:
I guess ultimately we want to both fuzz test this and let it run against a real Bitcoin Core node. But for now this will be sufficient :/

stevenroose replied:
I have only once dug into the fuzz code here, and I don't think it would be all too hard to do.

stevenroose commented:
Btw, do you agree with the Core naming of `HeaderAndShortId` instead of something like `CompactBlock`?

sgeisler commented:
Yeah, the name is a bit silly, but I prefer following the BIP/Core, less confusing imo than doing our own thing.

sgeisler reviewed (dismissed):
If you feel like adding some more tests/fuzzing feel free to do so. But I think this PR has reached a good enough state.

jrawsthorne commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,438 @@
+}
+
+/// Short transaction IDs are used to represent a transaction without sending a full 256-bit hash.
+#[derive(PartialEq, Eq, Clone, Default)]
```
Might as well derive copy and derive hash so ShortId can be used as a key in a map (I need this)

```suggestion
#[derive(PartialEq, Eq, Clone, Default, Hash, Copy)]
```

apoelstra reviewed (dismissed):
ACK, although because of the change to the network enum this is a breaking change.

RCasatta commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,438 @@
+    pub fn from_block(
+        block: &Block,
+        nonce: u64,
+        version: u32,
```
Just wondering if it would be a good idea to introduce a "BlockStandardVersion" enum, in this function would remove the version check and the `unreachable()!` macro calls

stevenroose replied:
Hmm, like this, the error only happens when creating one. At least we can decode newer versions from a byte stream and then notice that it has a version we don't understand or so..

stevenroose commented:
Rebased.

jrawsthorne commented on src/network/message_compact_blocks.rs:
```diff
@@ -0,0 +1,44 @@
+use util::bip152;
+
+/// sendcmpct message
+#[derive(PartialEq, Eq, Clone, Debug)]
```
Really small nit
```suggestion
#[derive(PartialEq, Eq, Clone, Debug, Copy)]
```

stevenroose replied:
ack

apoelstra reviewed (dismissed):
utack

justinmoon commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,438 @@
+
+    fn description(&self) -> &str {
+        "description() is deprecated; use Display"
+    }
```
Are either of these method implementations necessary? 

Nowhere else do we implement `Error::description` and `Error::cause` [defaults to `None`](https://doc.rust-lang.org/src/std/error.rs.html#151).

apoelstra replied:
No, they were needed when our MSRV was 1.22 (when this PR was opened :))

justinmoon commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,438 @@
+    /// > difference between the current index and the previous index, minus one.
+    /// > For example, a first index of 0 implies a real index of 0, a second
+    /// > index of 0 thereafter refers to a real index of 1, etc.
+    pub idx: u64,
```
Nit: this could be a u32 or maybe even a u16 (won't have 65535 transactions per block anytime soon)

sgeisler replied:
Agreed, especially since [Core uses a `uint16_t` too](https://github.com/bitcoin/bitcoin/blob/6a726cb534ed471010bf3f466732874b2a1b22e5/src/blockencodings.h#L71).

stevenroose replied:
fixed

sgeisler commented:
@apoelstra @stevenroose I'd like to get this into 0.27, it has been open for far too long and afaik the last iterations were only changing minor things. I hope the CI break is easy to fix.

apoelstra commented:
Yeah it looks super easy to fix

stevenroose commented:
Fixed nits and rebased. I couldn't test with 1.29 myself, let's see if CI could.

sgeisler reviewed (approved):
utACK 282facd3ded5708e5f039b55ce77587c9306966b

dr-orlovsky commented:
Needs rebase

vincenzopalazzo reviewed (commented):
utack https://github.com/rust-bitcoin/rust-bitcoin/commit/f06873c3ca5f11c3a5e8b3c1bcddd9d7d78bdbfd

This looks like that here it is missing only a rebase and fixed the CI error. that I'm not able to catch at the moment.

stevenroose commented:
Rebased.

dr-orlovsky commented on src/network/message_compact_blocks.rs:
```diff
@@ -0,0 +1,44 @@
+use util::bip152;
+
+/// sendcmpct message
+#[derive(PartialEq, Eq, Clone, Debug, Copy)]
```
With what we are putting into our guidelines (and what is in Rust guidelines) we do derive `PartialOrd, Ord, Hash` opportunistically. The same applies to the structures below. And, if I am not wrong, some of them may be `Copy`.

dr-orlovsky commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,441 @@
+use util::endian;
+
+/// A BIP-152 error
+#[derive(Clone, PartialEq, Eq, Debug)]
```
+ Copy, Ord's, Hash

dr-orlovsky commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,441 @@
+
+/// A [PrefilledTransaction] structure is used in [HeaderAndShortIds] to
+/// provide a list of a few transactions explicitly.
+#[derive(PartialEq, Eq, Clone, Debug)]
```
Plus `PartialOrd, Ord, Hash`

dr-orlovsky commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,441 @@
+    pub tx: Transaction,
+}
+
+impl convert::AsRef<Transaction> for PrefilledTransaction {
```
Nit: Isn't `AsRef` a part of prelude and we do not need to `use core::covert` or do `convert::` prefix here?

dr-orlovsky commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,441 @@
+}
+
+/// Short transaction IDs are used to represent a transaction without sending a full 256-bit hash.
+#[derive(PartialEq, Eq, Clone, Copy, Hash, Default)]
```
`PartialOrd, Ord`

dr-orlovsky commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,441 @@
+/// A [HeaderAndShortIds] structure is used to relay a block header, the short
+/// transactions IDs used for matching already-available transactions, and a
+/// select few transactions which we expect a peer may be missing.
+#[derive(PartialEq, Eq, Clone, Debug)]
```
`PartialOrd, Ord, Hash`

dr-orlovsky commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,441 @@
+
+/// A [BlockTransactionsRequest] structure is used to list transaction indexes
+/// in a block being requested.
+#[derive(PartialEq, Eq, Clone, Debug)]
```
`PartialOrd, Ord, Hash`

dr-orlovsky commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,441 @@
+    fn consensus_decode<D: io::Read>(mut d: D) -> Result<BlockTransactionsRequest, encode::Error> {
+        Ok(BlockTransactionsRequest {
+            block_hash: BlockHash::consensus_decode(&mut d)?,
+            indexes: {
```
Nit: Not sure whether it is a good style do inline computing here. At least we do not usually do that in other parts of this library

dr-orlovsky commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,441 @@
+
+/// A transaction index is requested that is out of range from the
+/// corresponding block.
+#[derive(Clone, PartialEq, Eq, Debug)]
```
`Copy, PartialOrd, Ord, Hash`

dr-orlovsky commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,441 @@
+
+/// A [BlockTransactions] structure is used to provide some of the transactions
+/// in a block, as requested.
+#[derive(PartialEq, Eq, Clone, Debug)]
```
`PartialOrd, Ord, Hash`

dr-orlovsky reviewed (commented):
I didn't use BIP-158, so just did a lite code review on the style and our guidelines and left few nits

stevenroose commented:
@dr-orlovsky Addressed all your derives, had to derive `PartialOrd, Ord, Hash` on `BlockHeader` (which tbh I think makes more sense than on most of the compact block types), I did that in a separate commit.

vincenzopalazzo reviewed (approved):
ack https://github.com/rust-bitcoin/rust-bitcoin/pull/249/commits/1d519df2b300fe5c5c31b5134ffce0a82a501914

dr-orlovsky reviewed (commented):
Concept ACK. I am not knowledgeable in compact filters to do a code vs standard review though :(

apoelstra commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,425 @@
+                            // strip witness for version 1
+                            let mut no_witness = tx.clone();
+                            no_witness.input.iter_mut().for_each(|i| i.witness.clear());
+                            no_witness
```
I know this PR is a million years old, but can you provide a citation/explanation for this witness-stripping behavior? I don't see it in Core or in BIP 152 or 158.

stevenroose replied:
It's been ages indeed, but it says that txs inside v1 should be like responses to MSG_TX and v2 should be like responses to MSG_WITNESS_TX:
https://github.com/bitcoin/bips/blob/master/bip-0152.mediawiki#specification-for-version-2

apoelstra commented:
Done review of 1d519df2b300fe5c5c31b5134ffce0a82a501914

ACK except for the thing about witness stripping.

I think this PR beats Carl's PSBT PR for "longest open PR" :)

tcharding commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,449 @@
+    }
+}
+
+impl hex::FromHex for ShortId {
```
Perhaps we should use `impl_bytes_newtype` to get this and the impls below?

tcharding commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,449 @@
+
+/// Short transaction IDs are used to represent a transaction without sending a full 256-bit hash.
+#[derive(PartialEq, Eq, Clone, Copy, Hash, Default, PartialOrd, Ord)]
+pub struct ShortId(pub [u8; 6]);
```
Should we have an `AsRef` implementation for `ShortId` as well?

0xB10C commented:
Nice, was looking for the BIP152 network messages in rust-bitcoin and was about to implement them myself. Will test against Bitcoin Core's network messages and review!

0xB10C commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,433 @@
+    ///  The blockhash of the block which the transactions being requested are in.
+    pub block_hash: BlockHash,
+    ///  The indexes of the transactions being requested in the block.
+    pub indexes: Vec<u64>,
```
I guess using `u64` is fine here though we'll probably not even going to get close to `u32::MAX` transactions in a block too.

tcharding commented:
Hey mate, this PR is currently marked as part of the 0.29 milestone, would you like to get this functionality in before the 0.29 release?

0xB10C commented:
I'm interested in seeing this in 0.29. I've rebased this branch on current `master`: https://github.com/0xB10C/rust-bitcoin/tree/compact-blocks - maybe this helps speed things up. I've been using  68760f50446553de085e2ed320a36e6ff5c1a071 to deserialize Bitcoin Core's compact block P2P messages for more than a month now without problems.

tcharding commented:
Legend @0xB10C, thanks. Perhaps if @stevenroose does not pick this up in a week or so you could create a PR and just mention that its Steven's work and it supersedes this one [0].  I can do it if you either don't have time or don't feel comfortable doing so.

I'm totally speaking for @stevenroose here, Steven you can go right ahead and slap me if I'm out of place :)


[0]: When doing so I typically add a `co-authored-by` tag to any commit if I make any non-trivial changes.

0xB10C commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,433 @@
+
+                let mut indexes = Vec::with_capacity(nb_indexes);
+                let mut last_index: u64 = 0;
+                for _ in 0..nb_indexes {
+                    let differential: VarInt = Decodable::consensus_decode(&mut d)?;
+                    last_index  = match last_index.checked_add(differential.0 + 1) {
+                        Some(r) => r,
+                        None => return Err(encode::Error::ParseFailed("block index overflow")),
+                    };
+                    indexes.push(last_index);
```
for reference: While adding `deserialize(serialize(BlockTransactionsRequest{..}))` test I noticed a subtle off-by-one error here. The increment in L321 (`differential.0 + 1`) is too early. This should only be incremented after we `indexes.push(last_index);`. Otherwise a differentially encoded `[0, 1, 0]` would become `[1, 3, 4]` when it should be `[0, 2, 3]`.

Fixed this in [TODO:].

tcharding commented:
Superseded by https://github.com/rust-bitcoin/rust-bitcoin/pull/1088

I've messaged @stevenroose via text also to make sure he is aware of this.
